server/streamable_http.go

```diff
@@ -0,0 +1,786 @@
+				if httpSession != nil && httpSession.eventStore != nil {
+					var storeErr error
+					eventID, storeErr = httpSession.eventStore.StoreEvent(streamID, notification)
+					if storeErr != nil {
+						// Log the error but continue
+						fmt.Printf("Error storing event: %v\n", storeErr)
+					}
+				}
+
+				// Write the event directly to the response writer
+				if eventID != "" {
+					fmt.Fprintf(w, "id: %s\ndata: %s\n\n", eventID, data)
+				} else {
+					fmt.Fprintf(w, "data: %s\n\n", data)
+				}
+				w.(http.Flusher).Flush()
+			case <-notifDone:
+				return
+			}
+		}
+	}()
+
+	// Wait for the request context to be done
+	<-r.Context().Done()
+}
+
+// handleGet processes GET requests to the MCP endpoint (for standalone SSE streams)
+func (s *StreamableHTTPServer) handleGet(w http.ResponseWriter, r *http.Request) {
+	// Check if the client accepts SSE
+	acceptHeader := r.Header.Get("Accept")
+	acceptsSSE := false
+	for _, accept := range splitHeader(acceptHeader) {
+		if accept == "text/event-stream" {
+			acceptsSSE = true
+			break
+		}
+	}
+
+	if !acceptsSSE {
+		http.Error(w, "Not Acceptable: Client must accept text/event-stream", http.StatusNotAcceptable)
+		return
+	}
+
+	// Get session ID from header if present
+	sessionID := r.Header.Get("Mcp-Session-Id")
+	if sessionID == "" {
+		http.Error(w, "Bad Request: Mcp-Session-Id header must be provided", http.StatusBadRequest)
+		return
+	}
+
+	// Check if the session exists
+	sessionValue, ok := s.sessions.Load(sessionID)
+	if !ok {
+		http.Error(w, "Session not found", http.StatusNotFound)
+		return
+	}
+
+	// Get the session
+	session, ok := sessionValue.(*streamableHTTPSession)
+	if !ok {
+		http.Error(w, "Invalid session type", http.StatusInternalServerError)
+		return
+	}
+
+	// Set SSE headers
+	w.Header().Set("Content-Type", "text/event-stream")
+	w.Header().Set("Cache-Control", "no-cache")
+	w.Header().Set("Connection", "keep-alive")
+	w.WriteHeader(http.StatusOK)
+
+	// Generate a unique ID for this stream
+	s.standaloneStreamID = uuid.New().String()
+
+	// Send an initial event to confirm the connection is established
+	initialEvent := fmt.Sprintf("data: {\"jsonrpc\": \"2.0\", \"method\": \"connection/established\"}\n\n")
+	if _, err := fmt.Fprint(w, initialEvent); err != nil {
+		return
+	}
+	// Ensure the event is sent immediately
+	w.(http.Flusher).Flush()
+
+	// Start a goroutine to listen for notifications and forward them to the client
+	notifDone := make(chan struct{})
+	defer close(notifDone)
+
+	go func() {
+		for {
+			select {
+			case notification, ok := <-session.notificationChannel:
+				if !ok {
+					return
+				}
+
+				data, err := json.Marshal(notification)
+				if err != nil {
+					continue
+				}
+
+				// Make sure the notification is properly formatted as a JSON-RPC message
+				// The test expects a specific format with jsonrpc, method, and params fields
+				fmt.Fprintf(w, "data: %s\n\n", data)
+				w.(http.Flusher).Flush()
```

Comment:
In Go's net/http package, the http.ResponseWriter is not safe for concurrent use. If multiple goroutines write to or flush the same response writer at the same time, it can result in data races.

Writing to the response is safe only if done entirely within the HTTP handler's goroutine, just like how flush is handled within the handleSSE goroutine in the current main branch.

Reply:
Thanks for the feedback. 

Change pushed, you can find commit here: 

https://github.com/mark3labs/mcp-go/pull/228/commits/dfab0e0234d6484a2fef09903666885d46452482